File: C++/20.Valid_Parentheses.cpp
class Solution {
	public:
		Solution() {
			std::ios_base::sync_with_stdio(false);
			std::cin.tie(NULL);
			std::cout.tie(NULL);
		}

		bool isValid(string s) {
			stack<char> myStack;
			int sLength = s.length();
			bool flag = true;
			if (sLength == 1 || s.at(0) == ')' || s.at(0) == '}' || s.at(0) == ']') {
				flag = false;
			} else {
				for (int i = 0; i < sLength; i++) {
					switch (s.at(i)) {
						case '(': 
							myStack.push('(');
							break;
						case '{': 
							myStack.push('{');
							break;
						case '[': 
							myStack.push('[');
							break;
						case ')':
							if (!myStack.empty() && myStack.top() == '(') {
								myStack.pop();
							} else {
								flag = false;
							} // if-else
							break;
						case '}':
							if (!myStack.empty() && myStack.top() == '{') {
								myStack.pop();
							} else {
								flag = false;
							} // if-else
							break;
						case ']':
							if (!myStack.empty() && myStack.top() == '[') {
								myStack.pop();
							} else {
								flag = false;
							} // if-else
							break;
					} // switch
				} // for
			} // if-else

			if (!myStack.empty()) flag = false;

			return flag;
		} // isValid
};

File: Python3/20.Valid_Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        '''
        Demonstrate stack implementation using LIST.
            -> st = []
        Functions:
            append() -> st.append('(')
            pop() -> st.pop(): LIFO, st.pop(3): delete fourth value
            len() -> len(st): size of the list 
        '''
        
        sLength = len(s)
        st = []
        flag = True
        
        if sLength < 2 or s[0] == ')' or s[0] == ']' or s[0] == '}' :
            flag = False
        else :
            for i in s:
                if i == '(': st.append(i)
                elif i == '{': st.append(i)
                elif i == '[': st.append(i)
                elif i == ')': 
                    if len(st) != 0 and st[-1] == '(': st.pop()
                    else: flag = False
                elif i == '}': 
                    if len(st) != 0 and st[-1] == '{': st.pop()
                    else: flag = False
                elif i == ']': 
                    if len(st) != 0 and st[-1] == '[': st.pop()
                    else: flag = False


        if len(st) != 0: flag = False
            
        
        return flag 
